`azurerm_spring_cloud_service` - support virtual network integration

njuCZ opened the pull request.

tombuildsstuff commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -64,6 +68,58 @@ func resourceArmSpringCloudService() *schema.Resource {
+							ValidateFunc: networkValidate.SubnetID,
+						},
+
+						"cidr": {
```
should this be `cidr_ranges`, since this isn't a single cidr range?

tombuildsstuff commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -64,6 +68,58 @@ func resourceArmSpringCloudService() *schema.Resource {
+							ForceNew: true,
+						},
+
+						"app_network_resource_group": {
```
can we not obtain this from the network id?

njuCZ replied:
above

tombuildsstuff commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -64,6 +68,58 @@ func resourceArmSpringCloudService() *schema.Resource {
+							},
+						},
+
+						"service_runtime_network_resource_group": {
```
can we not obtain this from the service runtime network id?

njuCZ replied:
above

tombuildsstuff commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -325,6 +383,11 @@ func resourceArmSpringCloudServiceRead(d *schema.ResourceData, meta interface{})
 			return fmt.Errorf("failure setting `trace`: %+v", err)
 		}
+		if network, err := flattenArmSpringCloudNetwork(ctx, loadBalancerClient, resp.Properties.NetworkProfile); err != nil {
+			return err
```
```suggestion
			return fmt.Errorf("flattening `network`: %+v", err)
```

tombuildsstuff commented on azurerm/internal/services/appplatform/tests/spring_cloud_service_resource_test.go:
```diff
@@ -252,6 +281,90 @@ resource "azurerm_spring_cloud_service" "test" {
+
+resource "azurerm_role_assignment" "test" {
+  scope                = azurerm_virtual_network.test.id
+  role_definition_name = "Owner"
```
presumably this can be Contributor?

bowen5 replied:
Unfortunately no. The target identity to be assigned is an Azure Spring Cloud Service owned one, which represents the service. And currently it requires Owner permission to inject the service instance into customer's VNet.

tombuildsstuff commented on website/docs/r/spring_cloud_service.html.markdown:
```diff
@@ -75,6 +77,20 @@ The following arguments are supported:
 
 The `config_server_git_setting` block supports the following:
 
+* `service_runtime_subnet_id` - (Required) Specifies the subnet id to host Azure Spring Cloud Service Runtime.
```
what does this mean in practice?

bowen5 replied:
"(Required) Specifies the full qualified resource id of an Azure Virtual Network Subnet, which will be used to host the Service Runtime components of the Azure Spring Cloud Service Instance to be created."

tombuildsstuff replied:
we avoid the phrase "Resource ID" since it's ambiguous between Azure and Terraform - so perhaps:

> * `service_runtime_subnet_id` - (Required) Specifies the ID of the Subnet where the Service Runtime components of the Spring Cloud Service will exist. Changing this forces a new resource to be created.

njuCZ replied:
done

tombuildsstuff commented on website/docs/r/spring_cloud_service.html.markdown:
```diff
@@ -75,6 +77,20 @@ The following arguments are supported:
 
+* `service_runtime_subnet_id` - (Required) Specifies the subnet id to host Azure Spring Cloud Service Runtime.
+
+* `app_subnet_id` - (Required) Specifies the subnet id to host Azure Spring Cloud Apps.
```
what does this mean in practice?

bowen5 replied:
"(Required) Specifies the full qualified resource id of an Azure Virtual Network Subnet, which will be used to host the Spring Boot microservices Applications of the Azure Spring Cloud Service Instance to be created."
How do you think?

tombuildsstuff replied:
(as above) we avoid the phrase Resource ID since it's ambiguous - but this could be:

> (Required) Specifies the ID of the Subnet which should host the Spring Boot Applications deployed in this Spring Cloud Service. Changing this forces a new resource to be created.

njuCZ replied:
done

tombuildsstuff commented on website/docs/r/spring_cloud_service.html.markdown:
```diff
@@ -75,6 +77,20 @@ The following arguments are supported:
+
+* `app_subnet_id` - (Required) Specifies the subnet id to host Azure Spring Cloud Apps.
+
+* `cidr` - (Required) Specifies the Spring Cloud service reserved CIDR.
```
what does this mean in practice?

bowen5 replied:
"(Required) Comma-separated list of IP address ranges in CIDR format. The IP ranges are reserved to host underlying Azure Spring Cloud infrastructure, which should be 3 at least /16 unused IP ranges, must not overlap with any Subnet IP ranges."
How do you think?

tombuildsstuff replied:
so then we should also add MinItems: 3 to the schema - @njuCZ can you update that? Perhaps this wants to be:

> (Required) A list of (at least 3) CIDR ranges (at least `/16`) which are used to host the Spring Cloud infrastructure, which must not overlap with any Subnet IP ranges.

tombuildsstuff commented on website/docs/r/spring_cloud_service.html.markdown:
```diff
@@ -137,6 +153,12 @@ The following attributes are exported:
+
+The `network` block exports the following:
+
+* `load_balancer_ip` - The Load Balancer IP for Azure Spring Cloud Service Runtime.
```
what does this mean in practice?

bowen5 replied:
'app_endpoint_ip' - The IP address of customer's Spring Boot microservices Applications which deployed in the Azure Spring Cloud Service Instance.

How do you think

njuCZ replied:
after discussed with bowen5, I have deleted this field.

tombuildsstuff reviewed (changes requested):
hi @njuCZ 

Thanks for this PR.

I've taken a look through and left some comments inline, but it looks like we can consolidate the schema here - if we can work through those then we should be able to take another look.

Thanks!

njuCZ commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -361,6 +424,27 @@ func resourceArmSpringCloudServiceDelete(d *schema.ResourceData, meta interface{
+		network.ServiceRuntimeNetworkResourceGroup = utils.String(serviceRuntimeNetworkResourceGroup)
+	}
+	if appNetworkResourceGroup := v["app_network_resource_group"].(string); appNetworkResourceGroup != "" {
+		network.AppNetworkResourceGroup = utils.String(appNetworkResourceGroup)
```
The backend service will create some resources (for example: load balancer) in `AppNetworkResourceGroup` and `ServiceRuntimeNetworkResourceGroup`, which serves different purpose for spring cloud runtime and user's app. If user does not specify these two fields, the backend service will automatically create two resource group.

njuCZ commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -361,6 +424,27 @@ func resourceArmSpringCloudServiceDelete(d *schema.ResourceData, meta interface{
+	}
+	if serviceRuntimeNetworkResourceGroup := v["service_runtime_network_resource_group"].(string); serviceRuntimeNetworkResourceGroup != "" {
+		network.ServiceRuntimeNetworkResourceGroup = utils.String(serviceRuntimeNetworkResourceGroup)
+	}
```
above

njuCZ commented on azurerm/internal/services/appplatform/spring_cloud_service_resource.go:
```diff
@@ -698,3 +782,65 @@ func flattenArmSpringCloudTrace(input *appplatform.TraceProperties) []interface{
+	}
+
+	if len(serviceRuntimeNetworkResourceGroup) != 0 {
+		lbName := "kubernetes-internal"
```
After discussed with service team, this is an internal implementations and it might change, so I have removed these codes

njuCZ commented:
@tombuildsstuff I have updated this PR. About the field description, the service team has given more context about it. and we have removed the field of internal load balancer. Please have a look again

njuCZ commented:
it seems the CI is pending, I will close and reopen this PR to trigger CI again

njuCZ commented:
@tombuildsstuff Could you please have a look again?

tombuildsstuff reviewed (approved):
LGTM 👍

tombuildsstuff commented:
Tests pass:

<img width="340" alt="Screenshot 2020-09-24 at 12 32 31" src="https://user-images.githubusercontent.com/666005/94134222-05749d00-fe62-11ea-8536-248c480d2cf9.png">
